### PR TITLE
Factura Audit Fixes

### PR DESCRIPTION
**Pull request by arminvoid**

### CODE REVIEW

File: contracts/interfaces/IController.sol [hunk 1]
```diff
@@ -2,9 +2,9 @@ pragma solidity 0.5.17;
 
 import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 
-contract IController {
+interface IController {
```
Comment (arminvoid) [inline contracts/interfaces/IController.sol:5]: Issue ID: 21

File: contracts/peaks/Controller.sol [hunk 2]
```diff
@@ -22,6 +22,9 @@ contract Controller is OwnableProxy {
     // Reserved storage space to allow for layout changes in the future.
     uint256[20] private _gap;
 
+    event PeakAdded(address indexed peak);
+    event VaultAdded(address indexed token, address indexed vault);
```
Comment (arminvoid) [inline contracts/peaks/Controller.sol:26]: Issue ID: 16

File: contracts/peaks/Controller.sol [hunk 3]
```diff
@@ -74,13 +76,17 @@ contract Controller is OwnableProxy {
     // Privileged methods
 
     function addPeak(address peak) external onlyOwner {
+        require(!peaks[peak], "Peak is already added");
```
Comment (arminvoid) [inline contracts/peaks/Controller.sol:79]: Issue ID: 15

File: contracts/peaks/Controller.sol [hunk 4]
```diff
@@ -34,8 +37,7 @@ contract Controller is OwnableProxy {
         IVault vault = vaults[address(token)];
         uint b = token.balanceOf(address(this));
         if (b > 0) {
-            token.safeApprove(address(vault), 0);
-            token.safeApprove(address(vault), b);
+            token.approve(address(vault), b);
```
Comment (arminvoid) [inline contracts/peaks/Controller.sol:40]: Issue ID: 10

File: contracts/base/Core.sol [hunk 5]
```diff
@@ -90,15 +88,15 @@ contract Core is OwnableProxy, Initializable, ICore {
         external
         returns(uint)
     {
-        Peak memory peak = peaks[msg.sender];
+        Peak storage peak = peaks[msg.sender];
```
Comment (arminvoid) [inline contracts/base/Core.sol:91]: Issue ID: 5

File: contracts/base/Core.sol [hunk 6]
```diff
@@ -68,13 +66,13 @@ contract Core is OwnableProxy, Initializable, ICore {
...
-        dusd = _dusd;
-        stakeLPToken = _stakeLPToken;
-        oracle = _oracle;
         require(
             _redeemFactor <= FEE_PRECISION && _colBuffer <= FEE_PRECISION,
             "Incorrect upper bound for fee"
         );
+        dusd = _dusd;
+        stakeLPToken = _stakeLPToken;
+        oracle = _oracle;
```
Comment (arminvoid) [inline contracts/base/Core.sol:75]: Issue ID: 4

File: contracts/base/Core.sol [hunk 7]
```diff
@@ -47,10 +47,8 @@ contract Core is OwnableProxy, Initializable, ICore {
 
     event Mint(address indexed account, uint amount);
     event Redeem(address indexed account, uint amount);
-    event FeedUpdated(uint[] feed);
     event TokenWhiteListed(address indexed token);
     event PeakWhitelisted(address indexed peak);
-    event UpdateDeficitState(bool inDeficit);
```
Comment (arminvoid) [inline contracts/base/Core.sol:53]: Issue ID: 2

File: contracts/peaks/Controller.sol [hunk 8]
```diff
@@ -74,13 +76,17 @@ contract Controller is OwnableProxy {
...
 
     function addPeak(address peak) external onlyOwner {
+        require(!peaks[peak], "Peak is already added");
         require(Address.isContract(peak), "peak is !contract");
         peaks[peak] = true;
+        emit PeakAdded(peak);
     }
 
     function addVault(address token, address vault) external onlyOwner {
+        require(address(vaults[token]) == address(0x0), "vault is already added for token");
```
Comment (cleanunicorn) [inline contracts/peaks/Controller.sol:86]: Good catch! This is similar to the check when adding a new peak.

**Pull request merged**